### PR TITLE
Add authentication

### PR DESCRIPTION
**Pull request by laptou**
You can now log in and log out, though being logged in currently has no effect.

https://drive.google.com/file/d/1-TTFrJo7EYInggQx6T2uHge9S81q6qDM/view?usp=sharing

### CODE REVIEW

File: portfolio/src/main/client/src/components/controls/authentication.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  role: 'admin' | 'user';
+}
+
+interface LoggedOutResponse {
+  loginUri: string;
+}
+
+type State = LoggedInResponse | LoggedOutResponse | null;
+
+async function updateState(state: DataSignal<State>) {
```
Comment (jessicaslaughter) [inline portfolio/src/main/client/src/components/controls/authentication.ts]: add JSDoc comment for all methods in this file

File: portfolio/src/main/client/src/components/controls/authentication.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  const currentState = state.value;
+  if (currentState && 'logoutUri' in currentState) {
+    // we don't even need to open the window to log out, so don't
+    void fetch(currentState.logoutUri)
+      .then(() => updateState(state))
+      .catch((err) => console.error(err));
+  }
+}
+
+export const Authentication = () => {
```
Comment (jessicaslaughter) [inline portfolio/src/main/client/src/components/controls/authentication.ts:87]: add JSDoc comment for exported item

**Review by jessicaslaughter — CHANGES_REQUESTED** [review]

File: portfolio/src/main/client/webpack.config.js [hunk 3]
```diff
@@ -18,25 +18,31 @@ const StylelintPlugin = require('stylelint-webpack-plugin');
...
-    },
-    historyApiFallback: true,
+    proxy: [{
+      context: ['/api', '/_ah'],
+      target: 'http://localhost:8080',
+      autoRewrite: true,
+      // changeOrigin: true,
+    }],
+    host: '0.0.0.0',
+    allowedHosts: ['localhost', '127.0.0.1', 'penguin.linux.test'],
```
Comment (jpnickolas) [inline portfolio/src/main/client/webpack.config.js:46]: Remove penguin.linux.test. Seems like something that was used for local debugging and wouldn't apply to other folks.

File: portfolio/src/main/client/src/util/data.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  public set value(value: T) {
+    const oldValue = this._value;
+    this._value = value;
+
+    for (const handler of this._handlers) {
+      handler(value, oldValue);
+    }
+  }
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/util/data.ts]: Use of setters to have wide reaching side effects makes code hard to follow and bugs hard to locate. Setters and getters should have side effects that are local to the class and predictable. For a pattern like this, I would recommend using a verbose model (signal.updateValue(value) and signal.getValue()).

File: portfolio/src/main/client/src/util/data.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Registers a handler for this signal.
+   *
+   * @param handler The handler which will respond to future value changes on
+   * this signal.
+   * @returns A cleanup function which will unregister this handler when called.
+   */
+  public on(handler: Handler<T>): Cleanup {
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/util/data.ts]: Name this something clearer like `onDataChange`. Currently the name is vague.

Comment (jpnickolas) [reply portfolio/src/main/client/src/util/data.ts]: Also update `off`

File: portfolio/src/main/client/src/components/controls/authentication.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+      container.classList.toggle('loading', false);
+      container.innerHTML = '';
+      container.append(loggedInContent);
+      return;
+    }
+
+    throw new Error('unexpected state');
+  }
+
+  state.on(handle);
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/controls/authentication.ts]: DataSignal seems pretty heavy handed for something that'll only ever have one callback. Since you only ever update `state.value` in `updateState`, why not call `handle` at the end of `updateState`? If you need to add more callbacks in the future, I'd recommend adding a function to your return object like `onStateChange` to handle this instead of tying it to the state object itself.

File: portfolio/src/main/client/src/components/controls/authentication.ts [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+  window.addEventListener('auth-login', () => {
+    void updateState(state);
+  });
+
+  window.addEventListener('auth-logout', () => {
+    void updateState(state);
+  });
+
+  function handle(newState: State) {
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/controls/authentication.ts]: Vague name. This should be something closer to `loginStateChanged`

File: portfolio/src/main/client/src/components/controls/authentication.ts [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    if (newState === null) {
+      container.classList.toggle('loading', false);
+      container.innerHTML = '';
+      container.append(loadingContent);
+      return;
+    }
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/components/controls/authentication.ts]: Every if block here is identical. You could reformat this to instead have an early return/error, but you're mostly asserting that `newState` is in fact of type `State` with this code. This is something that Typescript does for you already, so you should be able to get rid of the error as well.

File: portfolio/src/main/client/src/util/data.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+ * A function that unregisters a @see {Handler}.
+ *
+ * @returns `true` if that handler was registered, `false` otherwise.
+ */
+type Cleanup = () => boolean;
+
+/**
+ * A data signal, an object which notifies subscribers when data is modified.
+ */
+export class DataSignal<T> {
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/util/data.ts]: I'm not convinced you need to introduce this helper, and it probably hinders readability more than it helps it given the relatively small scope of this project.

File: portfolio/src/main/client/src/util/html.ts [hunk 10]
```diff
@@ -1,10 +1,19 @@
...
  */
-export type Renderable = Arrunk<string | Node | null | undefined>;
+export type PrimitiveRenderable = Arrunk<string | Node | null | undefined>;
+
+/**
+ * A component renderable, which has a certain state and renders a primitive.
+ */
+export interface ComponentRenderable {
+  render(): string | Node | null;
+}
```
Comment (jpnickolas) [inline portfolio/src/main/client/src/util/html.ts]: In which situations would you need this Renderable interface? I imagine that in a follow up PR, you'll have code like:
```
const auth = Authentication();
return htmls`Foo bar ${auth} baz`;
```

Why not call render (or better yet just provide a `root` variable so it turns into:
```
const auth = Authentication();
return htmls`Foo bar ${auth.root} baz`;
```

Comment (jpnickolas) [reply portfolio/src/main/client/src/util/html.ts]: Basically, I think that hiding the calls to `render` to be further down the stack makes this code harder to read and predict.

File: portfolio/src/main/java/com/google/sps/servlets/users/LandingServlet.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+package com.google.sps.servlets.users;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that allows the page to know if the user has logged in. */
```
Comment (jpnickolas) [inline portfolio/src/main/java/com/google/sps/servlets/users/LandingServlet.java]: nit: it seems like UserService redirects to this landing page after a user logs in or logs out. You should mention this in the class's docstring.

**Review by jpnickolas — CHANGES_REQUESTED** [review]

File: portfolio/src/main/client/webpack.config.js (hunk 3 above)
Comment (laptou) [reply portfolio/src/main/client/webpack.config.js:46]: Why does that mean I need to remove it? 
- For other users, it won't impact the functionality in any way
- `penguin.linux.test` seems to be universal to all Chromebooks anyway.

File: portfolio/src/main/client/src/components/controls/authentication.ts (hunk 6 above)
Comment (laptou) [reply portfolio/src/main/client/src/components/controls/authentication.ts]: The idea is that potentially, I will want to respond to changes in the authentication state from all over my app. Anyway, I like the `DataSignal` abstraction even if there's only one callback because it makes it easy to write my code like a finite state machine.

File: portfolio/src/main/client/src/util/html.ts (hunk 10 above)
Comment (laptou) [reply portfolio/src/main/client/src/util/html.ts]: Okay, I've removed this.

File: portfolio/src/main/client/src/components/controls/authentication.ts (hunk 6 above)
Comment (laptou) [reply portfolio/src/main/client/src/components/controls/authentication.ts]: I guess that wasn't very obvious due to the fact that I did not `export` the authentication state. I have now fixed that. What do you think?

File: portfolio/src/main/client/webpack.config.js (hunk 3 above)
Comment (jpnickolas) [reply portfolio/src/main/client/webpack.config.js:46]: I assumed `penguin.linux.test` was a private domain you owned. Resolving.

File: portfolio/src/main/client/src/components/controls/authentication.ts (hunk 6 above)
Comment (jpnickolas) [reply portfolio/src/main/client/src/components/controls/authentication.ts]: https://en.wikipedia.org/wiki/KISS_principle

Instead you can dispatch a custom event with `window.dispatchEvent` that the rest of your code listens on. Rolling your own library here really only makes sense if you're going to have a large app with maybe 5+ DataSignal objects and enough developers that you want to isolate the global namespace.

Comment (laptou) [reply portfolio/src/main/client/src/components/controls/authentication.ts]: Makes sense.

**Review by jessicaslaughter — APPROVED** [review]

**Review by jpnickolas — APPROVED** [review]
